v3/src/models/history/history-service.ts

```diff
@@ -0,0 +1,33 @@
+import { IAnyStateTreeNode } from "mobx-state-tree"
+import { ICoreNotification } from "../../data-interactive/notification-core-types"
+import { ILogMessage } from "../../lib/log-message"
+
+export type INotify =
+  ICoreNotification |
+  (ICoreNotification | (() => Maybe<ICoreNotification>))[] |
+  (() => Maybe<ICoreNotification | ICoreNotification[]>)
+
+export interface IApplyModelChangeOptions {
+  log?: string | ILogMessage | (() => Maybe<string | ILogMessage>)
+  notify?: INotify
+  notifyTileId?: string
+  undoStringKey?: string
+  redoStringKey?: string
+}
+
+export interface IHistoryService {
+  handleApplyModelChange: (mstNode: IAnyStateTreeNode, options?: IApplyModelChangeOptions) => void
+}
+
+const gHistoryServiceRef: {ref?: IHistoryService} = {}
```

Comment:
What's the advantage to this over the simpler:
```suggestion
const gHistoryService: Maybe<IHistoryService>
```
Are you expecting to extend the object to more than just the `ref` property in the future?

Reply:
No you're right it is unnecessary. I was thinking some modules might import gHistoryService directly. So then it'd have to have a ref property so the value could be changed after import. But then I switched to using a `getHistoryService()` method which seems cleaner than `gHistoryService.ref`.  Anyhow this will be somewhat mute when this is moved into the ITileEnvironment.

Reply:
With the change to putting the history service in the tile environment, yes I could keep the global access available. However there aren't any clients that need that as far as I know. So it seems better to take that away, and then interface of the history service can be simplified.  In particular it shouldn't be necessary to pass in MST nodes to the methods just so they can find the right tree to work with. Additionally the history service implementation can assume the tile environment is always available.